lib/std/smartptrs.nim

```diff
@@ -0,0 +1,130 @@
+
+## C++11 like smart pointers. They always use the shared allocator.
+import std/isolation
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    deallocShared(p.val)
+
+proc `=copy`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+  ## Returns a unique pointer which has exclusive ownership of the value.
+  result.val = cast[ptr T](allocShared(sizeof(T)))
+  # thanks to '.nodestroy' we don't have to use allocShared0 here.
+  # This is compiled into a copyMem operation, no need for a sink
+  # here either.
+  result.val[] = extract val
+  # no destructor call for 'val: sink T' here either.
+
+template newUniquePtr*[T](val: T): UniquePtr[T] =
+  ## .. warning:: Using this template in a loop causes multiple evaluations of `value`.
+  newUniquePtr(isolate(val))
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: UniquePtr[T]): var T {.inline.} =
+  ## Returns a mutable view of the internal value of `p`.
+  when compileOption("boundChecks"):
+    assert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc `[]=`*[T](p: UniquePtr[T], val: T) {.inline.} =
+  (p[]) = val
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "nil"
+  else: "UniquePtr(" & $p.val[] & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## Shared ownership reference counting pointer.
+    val: ptr tuple[value: T, atomicCounter: int]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  if p.val != nil:
+    if atomicLoadN(addr p.val[].atomicCounter, AtomicConsume) == 0:
+      `=destroy`(p.val[])
+      deallocShared(p.val)
+    else:
+      discard atomicDec(p.val[].atomicCounter)
+
+proc `=copy`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
+  if src.val != nil:
+    discard atomicInc(src.val[].atomicCounter)
+  if dest.val != nil:
+    `=destroy`(dest)
+  dest.val = src.val
+
+proc newSharedPtr[T](val: sink Isolated[T]): SharedPtr[T] {.nodestroy.} =
+  ## Returns a shared pointer which shares
+  ## ownership of the object by reference counting.
+  result.val = cast[typeof(result.val)](allocShared(sizeof(result.val[])))
+  result.val.atomicCounter = 0
+  result.val.value = extract val
+
+template newSharedPtr*[T](val: T): SharedPtr[T] =
+  ## .. warning:: Using this template in a loop causes multiple evaluations of `value`.
+  newSharedPtr(isolate(val))
+
+proc isNil*[T](p: SharedPtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: SharedPtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    assert(p.val != nil, "deferencing nil shared pointer")
```

Comment:
```suggestion
    doAssert(p.val != nil, "deferencing nil shared pointer")
```
so that --assertions:off --boundscheck:on does the right thing

Reply:
I see thanks, well this feels like a hack, will raise `NilAccessDefect` instead.